File: files-app/src/main/java/l/files/operations/Progress.java
package l.files.operations;

import auto.parcel.AutoParcel;

@AutoParcel
public abstract class Progress {

    public static final Progress NONE = normalize(0, 0);

    Progress() {
    }

    public abstract long getTotal();

    public abstract long getProcessed();

    /**
     * Throws IllegalArgumentException if total < processed or negative
     */
    public static Progress create(long total, long processed) {
        if (total < 0) {
            throw new IllegalArgumentException("total=" + total);
        }
        if (processed < 0) {
            throw new IllegalArgumentException("processed=" + processed);
        }
        if (total < processed) {
            throw new IllegalArgumentException("total=" + total
                    + ", processed=" + processed);
        }
        return new AutoParcel_Progress(total, processed);
    }

    /**
     * If total is less than processed, set total and processed to have the
     * value of processed.
     */
    public static Progress normalize(long total, long processed) {
        if (total < processed) {
            return create(processed, processed);
        } else {
            return create(total, processed);
        }
    }

    public float getProcessedPercentage() {
        if (NONE == this) {
            return 1F;
        } else {
            return getProcessed() / (float) getTotal();
        }
    }

    public long getLeft() {
        return getTotal() - getProcessed();
    }

    public boolean isDone() {
        return getTotal() == getProcessed();
    }

}

File: files-app/src/main/kotlin/operations/Progress.kt
package l.files.operations

import kotlin.platform.platformStatic

data class Progress private(val total: Long, val processed: Long) {

    val processedPercentage: Float
        get() = if (NONE == this) {
            1F
        } else {
            processed.toFloat() / total.toFloat()
        }

    val left: Long get() = total - processed

    val isDone: Boolean get() = total == processed

    class object {

        val NONE = normalize(0, 0)

        /**
         * Throws IllegalArgumentException if total < processed or negative
         */
        platformStatic fun create(total: Long, processed: Long): Progress {
            if (total < 0) {
                throw IllegalArgumentException("total=" + total)
            }
            if (processed < 0) {
                throw IllegalArgumentException("processed=" + processed)
            }
            if (total < processed) {
                throw IllegalArgumentException("total=" + total + ", processed=" + processed)
            }
            return Progress(total, processed)
        }

        /**
         * If total is less than processed, set total and processed to have the
         * value of processed.
         */
        platformStatic fun normalize(total: Long, processed: Long) =
                if (total < processed) {
                    Progress(processed, processed)
                } else {
                    Progress(total, processed)
                }

    }
}
